### ISSUE TITLE
Error: end must be greater than start

### ISSUE THREAD

**Issue by lnewson** [Swatinem/rollup-plugin-dts#87]
Hi,

I was trying to use this in a fairly large, somewhat older project converted from JS recently and ran into what I believe is a bug. Here's a simple replication case:

index.d.ts
```
export interface Cache {
  destroy: () => void;
}
export declare const uniqueId: (prefix?: string) => string;
export declare const Cache: () => Cache;
```
rollup.config.js:
```
import dts from "rollup-plugin-dts";

const config = [
  {
    input: "./index.d.ts",
    output: [{ file: "./out.d.ts", format: "es" }],
    plugins: [dts()],
  },
];

export default config;
```
When trying to bundle the above, rollup will throw a `end must be greater than start` error. 

Interestingly, I found if I rearrange the above so the `uniqueId` function is after the `Cache` function then it'll work. However, given the size of the project that's not really a viable option for me personally.

So I just thought I'd log this in hopes I could get some help to get it resolved. I'm happy to look into it myself, I'm just not too sure where to start to be honest.

Thanks!

**Comment by lnewson** [comment]
I found https://github.com/Swatinem/rollup-plugin-dts/blob/master/src/Transformer.ts#L118, so it seems as though this might just be something that isn't supported?

**Comment by lnewson** [comment]
Here's another reproduction case if it's useful?
```
export interface Cache {
  add: (info: CacheInfo) => boolean;
  destroy: () => void;
}
export interface CacheInfo {
  id: number;
}
export declare const Cache: () => Cache;
```

Basically in most cases for code style, all imports are declared at the top of the file, so it's fairly rare the interface and variable will be next to each other.

**Comment by lnewson** [comment]
I also tried the following which will compile fine, but it drops the `declare const` statement in the output so the types generated are incorrect as the `Cache` type is normally merged.

Input:
```
interface Cache {
  add: (info: CacheInfo) => boolean;
  destroy: () => void;
}
interface CacheInfo {
  id: number;
}
declare const Cache: () => Cache;
export { Cache, CacheInfo };
```

Output:
```
interface Cache {
  add: (info: CacheInfo) => boolean;
  destroy: () => void;
}
interface CacheInfo {
  id: number;
}

export { Cache, CacheInfo };
```

**Comment by Swatinem** [comment]
> I found https://github.com/Swatinem/rollup-plugin-dts/blob/master/src/Transformer.ts#L118, so it seems as though this might just be something that isn't supported?

Yes, this is indeed the problem.

The thing is that rollup is built around the assumption that in JS one "name" refers to a single contiguous range of code. Which is just not the case in TS, since you can have overrides, like the comment says, and you can define both a *type* (your `interface`) and a *value* (your `declare const`) with the same name, and they can cover non-contiguous ranges of your code.

I can think of one rather straight forward way to deal with that:
Parse once, see if we have any non-contiguous pieces, re-order them to be next to each other, parse again. We can surely skip the second parse if there is nothing to reorder.

In either case, the current error message is super bad, and it would have been a good idea to error our sooner.

**Comment by Swatinem** [comment]
Just published `1.4.2` with the proposed approach. Can you verify with that?

**Comment by lnewson** [comment]
@Swatinem geez, you were extremely quick getting that out. I can confirm the project working I was testing now compiles fine and at a glance I can't see anything missing. I'll have a better look tomorrow and if I find anything I'll let you know and come up with a replication case. Thanks heaps for the quick turnaround on the fix! 👍
